lib/Dialect/SPIRV/SPIRVOps.cpp

```diff
@@ -1058,6 +1058,83 @@ OpFoldResult spirv::CompositeExtractOp::fold(ArrayRef<Attribute> operands) {
   return extractCompositeElement(operands[0], indexVector);
 }
 
+//===----------------------------------------------------------------------===//
+// spv.CompositeInsert
+//===----------------------------------------------------------------------===//
+
+static ParseResult parseCompositeInsertOp(OpAsmParser &parser,
+                                          OperationState &state) {
+  OpAsmParser::OperandType objectInfo, compositeInfo;
+  Type objectType, compositeType;
+  Attribute indicesAttr;
+  auto loc = parser.getCurrentLocation();
+
+  if (parser.parseOperand(objectInfo) || parser.parseComma() ||
+      parser.parseOperand(compositeInfo) ||
+      parser.parseAttribute(indicesAttr, kIndicesAttrName, state.attributes) ||
+      parser.parseColonType(objectType) ||
+      parser.parseKeywordType("into", compositeType) ||
+      parser.resolveOperands({objectInfo, compositeInfo},
+                             {objectType, compositeType}, loc,
+                             state.operands)) {
+    return failure();
+  }
+
+  state.addTypes(compositeType);
+  return success();
+}
+
+static LogicalResult verify(spirv::CompositeInsertOp compositeInsertOp) {
+  auto indicesArrayAttr = compositeInsertOp.indices().dyn_cast<ArrayAttr>();
+  if (!indicesArrayAttr.size()) {
+    return compositeInsertOp.emitOpError(
+        "expected at least one index for spv.CompositeInsert");
+  }
+
+  // TODO: `spv.CompositeInsert` and `spv.CompositeExtrat` share the same logic
+  // to extract an object from the given composite by given indices, should we
+  // merge it?
+  int32_t index;
+  auto objectType = compositeInsertOp.composite()->getType();
+  for (auto indexAttr : indicesArrayAttr) {
+    index = indexAttr.dyn_cast<IntegerAttr>().getInt();
+    if (auto cType = objectType.dyn_cast<spirv::CompositeType>()) {
+      if (index < 0 || static_cast<uint64_t>(index) >= cType.getNumElements()) {
+        return compositeInsertOp.emitOpError("index ")
+               << index << " out of bounds for " << objectType;
+      }
+      objectType = cType.getElementType(index);
+    } else {
+      return compositeInsertOp.emitError(
+                 "cannot extract from non-composite type ")
+             << objectType << " with index " << index;
+    }
+  }
+
+  if (objectType != compositeInsertOp.object()->getType()) {
+    return compositeInsertOp.emitOpError("invalid object type: expected ")
+           << objectType << " but provided "
+           << compositeInsertOp.object()->getType();
+  }
+
+  if (compositeInsertOp.composite()->getType() != compositeInsertOp.getType()) {
+    return compositeInsertOp.emitOpError("invalid result type: expected ")
```

Comment:
"result type should be the same as the composite operand type, but found " << ... << " vs " << ... ?